### ISSUE TITLE
Make Comparison with OpenAPI more in-depth

### ISSUE THREAD

**Issue by bobergj** [fern-api/fern#1134]
I find the project overall very interesting. I wanted to build something like this yesterday.

The [Comparison with OpenAPI](https://www.buildwithfern.com/docs/comparison-with-openapi) needs a more in-depth analysis in order to convince users why we need yet another specification standard™. 
For example: concrete examples of openapi specification-level issues that make code generation problematic.

It's pretty clear why I would want to have your code generation, with the AST and intermediate model approach, rather than https://github.com/OpenAPITools/openapi-generator and its templates.

In other words, explain why the problem cannot be solved with the combination of:
* a OpenAPI 3.1 linter that detects specification constructs that inhibit code generation
* a new code generation tool

rather than also introducing a new schema specification language.

**Comment by dannysheridan** [comment]
Thanks for taking the time to share your perspective! The AST is definitely a key part of that answer/story. Work to do on our end to better put words to why we need _another_ specification standard.

**Comment by zachkirsch** [comment]
Hey @bobergj, +1 thanks for the feedback. I just pushed up https://github.com/fern-api/fern/pull/1159 which adds some better examples and a bit more detail on our comparison with OpenAPI + the existing generators. Take a look at the [new docs page](https://docs.buildwithfern.com/comparison-with-openapi)!

To poke at the premise of your question, I don't think it's vital to answer the question "why can't this be done with OpenAPI extensions, a linter, and new generators?" Even if it _were_ possible, it would undermine one of our main values: making it dead simple to define APIs that generate great code. Maybe we'd still end up at great code, but at the expense of a good user experience.

I think also valuable to emphasize our compatibility with OpenAPI to Fern (our Fern->OpenAPI converter is live and our OpenAPI->Fern converter is in beta). With that conversion option, I could see Fern being used side-by-side with OpenAPI, each being used for their designated use cases.

**Comment by bobergj** [comment]
Hey, I had a look at the new doc page, and it's definitely an improvement! One minor nitpick: it would be more fair if also openapi examples where written in the openapi `yaml` serialization, rather than json.

> To poke at the premise of your question, I don't think it's vital to answer the question "why can't this be done with OpenAPI extensions, a linter, and new generators?" Even if it were possible, it would undermine one of our main values: making it dead simple to define APIs that generate great code. Maybe we'd still end up at great code, but at the expense of a good user experience.

Now I am gonna move the goalposts a bit.
Having looked at your new examples, sure, I don't want to write openapi ever again. Your schema language is better. But: I think you aren't taking it far enough. It still wouldn't be pleasant to author my api specs with that. One word: yaml. If you are gonna make a new schema description language from scratch, why yaml? I know the pros and cons of yaml have been debated to death. In this case though, the issue with yaml isn't about syntax (https://noyaml.com), its that it isn't expressive enough.

There are no variable declarations. For example, let's say the all the discriminated enums in my API have the discriminator ``__type``. If I was using a "proper" dsl or mainstream programming language I could specify that once with:
```
let myApiTypeDiscriminator = "__type"
```
and then just refer use this variable in the rest of my spec.

There are no parameterized construct in yaml. In your example, clearly `discriminant` is a property of `union`, but it's declared on the same level as enum:
```
Animal:
  discriminant: pet_type
  union:
    dog: Dog
    cat: Cat
```
This isn't great for discoverability / autocompletion.
Again, if I was using "proper" dsl or mainstream programming language I could perhaps write:
```
object Animal {
  union(discriminator: myApiTypeDiscriminator) {
    dog: Dog
    cat: Cat
  }
}
```

This idea obviously isn't new (graphql, grpc etc). Recently I've been has extensively authoring CI workflows in yaml. CircleCI yaml is pretty similar to github actions: https://circleci.com/docs/sample-config. If I did it again, I wouldn't write the yaml directly. I would use their new typescript SDK: https://circleci.com/blog/config-sdk/ which eventually outputs the same yaml.

**Comment by bobergj** [comment]
The same idea for json schema: https://github.com/zweifisch/kontur

**Comment by zachkirsch** [comment]
It's almost like you've been eavesdropping on our conversations! Couldn't agree more - for lots of devs, YAML is definitely not the right choice. For now, we've chosen it as a proof of concept and it won't be too difficult to add higher fidelity options later. Some things we're considering:
- DSL (like you brought up), e.g. proto. Agreed that the design of it is tricky and you still have to teach a new format to users.
- UI (like Stoplight w/ OpenAPI). For a lot of users, this is the easier and most preferred.
- TypeScript (like you brought up), e.g. Circle, CDK.

There's a few benefits of _not_ using TypeScript (or another language) to define your API:
1. At larger companies, it's really nice to be able to look at another team's API spec to understand their API. In a simple format (YAML, DSL), this is pretty easy if you know the spec. Once you allow arbitrary typescript, it's easy for teams to add a lot of complexity in the name of DRY that makes the API definition harder to grok.
2. It makes it impossible to write a language server. For example, in YAML or a DSL, we could write a language server for custom logic (e.g. warn that `snake_case` type names  should be `PascalCase`). In TypeScript, we can't show inline feedback like that in the IDE.

**Comment by dsinghvi** [comment]
Closing (docs now live in another repo https://github.com/fern-api/docs and we now accept OpenAPI as input)
